### PR TITLE
bpf: nat: skip snat when using iptables masquerade

### LINKED ISSUE #36572: If pods establish more external connections than the value specified by `--bpf-nat-global-max`, it causes a TCP Reset in SNAT behavior for `cil-to-netdev`
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.16.4 and lower than v1.17.0

### What happened?

When using `KubeProxyReplacement: Strict` with `Masquerading: IPTables [IPv4: Enabled]`, the source address is already nat-ed by iptables. However, `cil-to-netdev` performs an additional unnecessary SNAT behavior. It also unnecessarily occupies entry in the SNAT_MAPPING map. And if it exceeds SNAT_COLLISION_RETRIES, it leads to connection failures.

https://github.com/cilium/cilium/blob/daea10fa8ecf00c6248dc22df9b87937e039d431/bpf/lib/nat.h#L232-L243

### How can we reproduce the issue?

To easily reproduce the issue, you can reduce `--bpf-nat-global-max` to 1024 and send multiple connection from a pod to external packets. This should trigger the issue immediately.
```
fortio load -qps 1000 -c 1000 -t 100s -allow-initial-erros -keepalive=false http://{EXTERNAL_IP}
```

### Cilium Version

cilium-cli: v0.16.15 compiled with go1.22.5 on darwin/arm64
cilium image (default): v1.16.0
cilium image (stable): v1.16.4
cilium image (running): v1.16.4

### Kernel Version

Linux gyu-test-1 5.15.0-125-generic #135-Ubuntu SMP Fri Sep 27 13:53:58 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.31.2
Kustomize Version: v5.4.2
Server Version: v1.26.4

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell

```

### Anything else?

_No response_

### Cilium Users Document

- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by gyutaeb**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

When using iptables masquerade, the source address is already nat-ed by iptables. So, we should skip snat when using iptables masquerade. If we don't skip snat, the source address is nat-ed twice. It also unnecessarily occupies entry in the SNAT_MAPPING map. And if it exceeds SNAT_COLLISION_RETRIES, it leads to connection failures.

Fixes: #36572

### CODE REVIEW

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
>It is no longer correct to check the NAT_MIN_EGRESS condition. This is because target->from_local_endpoint is now being checked. In the case of !target->from_local_endpoint, source NAT is not required,

I don't follow. Can you please elaborate *much* more on the suggested change. **Why** is source NAT not required in the case you're describing? Also please add this in-depth explanation to the patch description, not just the PR.

**Comment by gyutaeb** [comment]
@julianwiedmann @gentoo-root Thank you for review! When using iptables masquerade instead of bpf masquerade, SNAT is not required. I plan to enhance the `cDefinesMap` and explanation of suggested change. I will re-request after that.

**Comment by gyutaeb** [comment]
@julianwiedmann @gentoo-root @ti-mo Thank you for your patience. The previous changes had some ambiguities. To clarify the bug and the fix, I added `cDefinesMap["ENABLE_IPTABLES_MASQUERADE_IPV4"]` and `cDefinesMap["ENABLE_IPTABLES_MASQUERADE_IPV6"]` to explicitly handle SNAT accordingly. Additionally, I provided a clear and detailed explanation of the #36572 and the commit description.
The issue is that when using iptables masquerade, the source address is already nat-ed by iptables. So, we should skip SNAT when using iptables masquerade. If we don't skip SNAT, the source address is nat-ed twice. It also unnecessarily occupies entry in the `SNAT_MAPPING` map. And if it exceeds `SNAT_COLLISION_RETRIES`, it leads to connection failures due to `DROP_NAT_NO_MAPPING`.
This patch ensures that SNAT is skipped when using iptables masquerade. It prevents duplicate SNAT, avoids unnecessary use of the `SNAT_MAPING` map, and mitigates connection failures caused by packet drops. Additionally, by eliminating the complexity caused by duplicate SNAT, it reduces troubleshooting time and clarifies the masquerade process.
The new define was named based on the code reference below.
https://github.com/cilium/cilium/blob/5a4fdc967a51146c15b65e5401ecb5d7decdd8d2/pkg/option/config.go#L2416-L2426

**Comment by gyutaeb** [comment]
*Update Branch(Rebase)

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
I don't see how this can be correct for the case of using BPF Nodeport without BPF masquerade. Please elaborate a lot more.

We're **intentionally** tracking connections in the SNAT engine for this case, and applying SNAT as needed to avoid conflicts between [LB-to-remote-backend connections](https://github.com/cilium/cilium/blob/d2c4e0f7baa6080bd78b4ae454e1c21de88e61ab/bpf/lib/nodeport.h#L2453) and [host-originating / iptables-masqueraded connections](https://github.com/cilium/cilium/blob/d2c4e0f7baa6080bd78b4ae454e1c21de88e61ab/bpf/lib/nodeport_egress.h#L347).

**Comment by gyutaeb** [comment]
> I don't see how this can be correct for the case of using BPF Nodeport without BPF masquerade. Please elaborate a lot more.
> 
> We're **intentionally** tracking connections in the SNAT engine for this case, and applying SNAT as needed to avoid conflicts between [LB-to-remote-backend connections](https://github.com/cilium/cilium/blob/d2c4e0f7baa6080bd78b4ae454e1c21de88e61ab/bpf/lib/nodeport.h#L2453) and [host-originating / iptables-masqueraded connections](https://github.com/cilium/cilium/blob/d2c4e0f7baa6080bd78b4ae454e1c21de88e61ab/bpf/lib/nodeport_egress.h#L347).

@julianwiedmann Thank you for your detailed explanation. I agree with your opinion, After reviewing the case you explained, I realized that skipping SNAT in iptables-masquerade cannot prevent the conflict with LB-to-remote-backend.
```
flow1: pod -> node:nodeport(32000) >>>> SNAT engine >>>> node:natedPort(34000) -> remote-backend
flow2: host -> external >>>> iptables-masquerade >>>> host:natedPort(34000) -> external
```
I hadn't considered that scenario, so thank you for pointing it out. This PR should not be merged. To elaborate, during my attempts to reproduce this issue, I observed a rare case where the SNAT engine performed SNAT on an egress packet, but the reverse NAT for the response packet failed. This occurred when the `--bpf-nat-global-max` limit was reached. I tried skipping SNAT in order to resolve this issue, but it's clear now that skipping SNAT is not the right approach. Instead, I will focus on analyzing why reverse NAT occasionally fails. If you have encountered or received reports about a similar issue, I'd greatly appreciate it if you could share them.
In summary, this PR should not be merged. I will focus on the reverse NAT issue and submit a new PR with an appropriate fix.

**Comment by julianwiedmann** [comment]
> In summary, this PR should not be merged. I will focus on the reverse NAT issue and submit a new PR with an appropriate fix.

Thank you for the feedback! Looking forward to what you come up with :).

I'll flip this PR to `draft` then for now.

**Comment by gyutaeb** [comment]
> I observed a rare case where the SNAT engine performed SNAT on an egress packet, but the reverse NAT for the response packet failed.

@julianwiedmann Hello, I'd like to share the root cause of the issue that caused REV NAT to fail, as mentioned earlier. This issue was mitigated by #35304, which introduced the idea of recreating evicted reverse entries during NAT using the otuple. I think this is a fantastic approach!
While #35304 has significantly reduced the issue, there are still rare cases where the reverse entry is evicted when `snat_v4_rev_nat_handle_mapping()` attempts to find it.
To address this, I'd like to propose an alternative: if `snat_v4_rev_nat_handle_mapping()` cannot find the reverse entry, it could attempt to recreate it using the original entry(assuming the original entry has not been deleted). What are your thoughts on this approach?

**Comment by julianwiedmann** [comment]
> > I observed a rare case where the SNAT engine performed SNAT on an egress packet, but the reverse NAT for the response packet failed.
> 
> @julianwiedmann Hello, I'd like to share the root cause of the issue that caused REV NAT to fail, as mentioned earlier. This issue was mitigated by #35304, which introduced the idea of recreating evicted reverse entries during NAT, using the otuple. I think this is a fantastic approach! While #35304 has significantly reduced the issue, there are still rare cases where the reverse entry is evicted when `snat_v4_rev_nat_handle_mapping()` attempts to find it. To address this, I'd like to propose an alternative: if `snat_v4_rev_nat_handle_mapping()` cannot find the reverse entry, it could attempt to recreate it using the original entry(assuming the original entry has not been deleted). What are your thoughts on this approach?

Sorry, I don't think I'll have the cycles to co-design a solution for the problem you encountered. But maybe you can connect with @sugangli and discuss this in more detail - hack up a minimal reproducer that demonstrates the problem etc?

**Comment by sugangli** [comment]
Hi @gyutaeb,  

IIUC,  retrieving the original entry from the rev nat path could be a bit tricky. Could you give an example on how it could be done? Feel free to tag me in the new PR once you are ready.

**Comment by gyutaeb** [comment]
Hi, @sugangli 
I will reach out to you once the code is ready. I'm excited to collaborate with you. Currently, I am working on a PR related to #31213, and it will take a few more weeks to complete. Thank you for your support!
